core/cpuid.c

```diff
@@ -114,3 +131,234 @@ bool cpuid_host_has_feature_uncached(uint32_t feature_key)
+    hax_log(HAX_LOGI, "feature_8000_0001_ecx: %08lx, feature_8000_0001_edx: %08lx"
+            "\n", cpuid->feature_8000_0001_ecx, cpuid->feature_8000_0001_edx);
+
+    for (int i = 0; i < count; ++i) {
+        cpuid_set_guest_feature[i](cpuid, cpuid_info);
+    }
+
+    hax_log(HAX_LOGI, "%s: after:\n", __func__);
+    hax_log(HAX_LOGI, "feature_1_ecx: %08lx, feature_1_edx: %08lx\n",
+            cpuid->feature_1_ecx, cpuid->feature_1_edx);
+    hax_log(HAX_LOGI, "feature_8000_0001_ecx: %08lx, feature_8000_0001_edx: %08lx"
+            "\n", cpuid->feature_8000_0001_ecx, cpuid->feature_8000_0001_edx);
+}
+
+static hax_cpuid_entry * find_cpuid_entry(hax_cpuid *cpuid_info,
+                                          uint32_t function, uint32_t index)
+{
+    int i;
+    hax_cpuid_entry *entry, *found = NULL;
+
+    for (i = 0; i < cpuid_info->total; ++i) {
+        entry = &cpuid_info->entries[i];
+        if (entry->function == function && entry->index == index) {
+            found = entry;
+            break;
+        }
+    }
+
+    return found;
+}
+
+static void cpuid_set_0000_0001(cpuid_t *cpuid, hax_cpuid *cpuid_info)
+{
+    const uint32_t kFunction = 0x01;
+    hax_cpuid_entry *entry;
+
+    entry = find_cpuid_entry(cpuid_info, kFunction, 0);
+    if (entry == NULL)
+        return;
+
+    hax_log(HAX_LOGI, "%s: function: %08lx, index: %lu, flags: %08lx\n",
+            __func__, entry->function, entry->index, entry->flags);
+    hax_log(HAX_LOGI, "%s: eax: %08lx, ebx: %08lx, ecx: %08lx, edx: %08lx\n",
+            __func__, entry->eax, entry->ebx, entry->ecx, entry->edx);
+
+    cpuid->feature_1_ecx = entry->ecx;
+    cpuid->feature_1_edx = entry->edx;
+
+    // Filter the unsupported features
+    cpuid->feature_1_ecx &= cache.host_supported.feature_1_ecx &
+                            cache.hax_supported.feature_1_ecx;
+    cpuid->feature_1_edx &= cache.host_supported.feature_1_edx &
+                            cache.hax_supported.feature_1_edx;
+
+    // Set fixed supported features
+    cpuid_set_fixed_features(cpuid);
+
+    if (entry->ecx != cpuid->feature_1_ecx ||
+        entry->edx != cpuid->feature_1_edx) {
+        hax_log(HAX_LOGW, "%s: filtered or unchanged flags: ecx: %08lx, "
+                "edx: %08lx\n", __func__, entry->ecx ^ cpuid->feature_1_ecx,
+                entry->edx ^ cpuid->feature_1_edx);
+    }
+}
+
+static void cpuid_set_8000_0001(cpuid_t *cpuid, hax_cpuid *cpuid_info)
+{
+    const uint32_t kFunction = 0x80000001;
+    hax_cpuid_entry *entry;
+
+    entry = find_cpuid_entry(cpuid_info, kFunction, 0);
+    if (entry == NULL)
+        return;
+
+    hax_log(HAX_LOGI, "%s: function: %08lx, index: %lu, flags: %08lx\n",
+            __func__, entry->function, entry->index, entry->flags);
+    hax_log(HAX_LOGI, "%s: eax: %08lx, ebx: %08lx, ecx: %08lx, edx: %08lx\n",
+            __func__, entry->eax, entry->ebx, entry->ecx, entry->edx);
+
+    cpuid->feature_8000_0001_edx = entry->edx;
```

Comment:
ecx is not merged. CUrrently supported by haxm bits are zero but this will change in future.

Reply:
You are right. This version is based on the current implementation and this feature will be easy to extend in future.

Reply:
We will update the implementation of CPUID and include a hash table for feature set, which will add ecx or other register settings. It will not change the HAX_VCPU_IOCTL_SET_CPUID interface. The patch will be ready once the current release completes.

Reply:
A hash table for searching entries by find_cpuid_entry? There are currently only two codes with such flags, a vector is sufficient for this task.

Reply:
A hash table for storing in hax_cpuid_t for future extensions although a vector is sufficient currently.